### ISSUE TITLE
no such file to load -- net/ldap 

### ISSUE THREAD

**Issue by huckabeec** [voxpupuli/puppet-ldapquery#37]
- Puppet: 7.23
- Ruby:  
```console
# /opt/puppetlabs/puppet/bin/ruby --version
ruby 2.7.7p221 (2022-11-24 revision 168ec2b1e5) [x86_64-linux]
```

- Distribution: foreman
- Module version:  Pulled from github

Getting this error:
```
Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Function Call, Could not autoload puppet/parser/functions/ldapquery: no such file to load -- net/ldap 
```

```console
# /opt/puppetlabs/puppet/bin/gem list  | grep net-ldap
net-ldap (0.17.1)
```

So is it hardwired somewhere where it can't use the latest version or is something else amiss ?

**Comment by huckabeec** [comment]
gem list is from the puppetserver; was trying to test the module to see if it could do what we need but got stuck on the error.

**Comment by ekohl** [comment]
Have you tried a manual way? I think `echo "require 'net/ldap'" | /opt/puppetlabs/puppet/bin/ruby` should work.

**Comment by huckabeec** [comment]
`# echo "require 'net/ldap'" | /opt/puppetlabs/puppet/bin/ruby
# `

**Comment by ekohl** [comment]
Just a thought: have you perhaps not restarted puppetserver after installing the gem?

**Comment by huckabeec** [comment]
Restarted puppetserver with no change.  It has to be something I'm doing wrong I just don't know what yet.

**Comment by ekohl** [comment]
Is there more information in puppetserver's log? Sometimes there's a backtrace.

**Comment by ekohl** [comment]
That doesn't really tell me anything so I'm out of ideas now.

**Comment by huckabeec** [comment]
I'll keep banging on it and post of I figure it out.

**Comment by huckabeec** [comment]
Not the solution to my error but another weird thing.  Created a test manifest:
```
$grpattr = [ 
  'memberuid'
]
  
$member_query = '(&(cn=mac-admin)(objectclass=posixgroup))'
$member_results = ldapquery($member_query, $grpattr);
$member_results.each |$g| { 
  any2array($g['memberuid']).each|$k| {
    warning($k)
  }
} 
```
Running that with `puppet apply test.pp -d > out 2>&1`

`ldapquery(): Searching  for ["memberuid"] using (&(cn=mac-admin)(objectclass=posixgroup)) took 0.321 seconds and returned 0 results`

Note how the basedn is missing after the 'Searching', despite the fact that the ldapbase is defined in puppet.conf on the master.

**Comment by smortex** [comment]
@huckabeec  Isn't the net-ldap gem supposed to be installed on the puppetserver side instead of the puppet agent?

Something like `sudo puppetserver gem install ...`?

**Comment by huckabeec** [comment]
> @huckabeec Isn't the net-ldap gem supposed to be installed on the puppetserver side instead of the puppet agent?
> 
> Something like `sudo puppetserver gem list` / `sudo puppetserver gem install ...`?

It was, double checking now that it is still there.

**Comment by huckabeec** [comment]
```
# /opt/puppetlabs/bin/puppetserver gem list |grep net-ldap
net-ldap (0.17.1)
```
I got the gem installed in the right place but still getting no results and no basedn in the debug log using the test manifest I mentioned above.

**Comment by smortex** [comment]
`/opt/puppetlabs/puppet/bin/ruby` is the AIO version of ruby (MRI) used by the agent.  The PuppetServer use jruby, I think you can try `curl -Lks 'https://git.io/rg-ssl' | puppetserver ruby` (assuming this thing is what I think it is)…

**Comment by huckabeec** [comment]
```
# curl -Lks 'https://git.io/rg-ssl'|/opt/puppetlabs/bin/puppetserver ruby
!!! WARNING !!!
Ruby 2.6.8 has reached end-of-life, and is unsupported.
This script may not work.

Here's your Ruby and OpenSSL environment:

Ruby:           2.6.8p0 (2022-10-24 revision 67951) [java]
RubyGems:       3.2.33
Bundler:        2.2.33
Compiled with:  JRuby-OpenSSL 0.14.0
Loaded version: JRuby-OpenSSL 0.14.0
SSL_CERT_FILE:  /etc/pki/tls/certs/ca-bundle.crt
SSL_CERT_DIR:   /etc/ssl/certs

With that out of the way, let's see if you can connect to rubygems.org...

Bundler connection to rubygems.org:       failed  ❌  (SSL SSLContext not available)
RubyGems connection to rubygems.org:      failed  ❌  (SSL SSLContext not available (https://rubygems.org))
Ruby net/http connection to rubygems.org: failed  ❌

Unfortunately, this Ruby can't connect to rubygems.org. 😡
Even worse, we're not sure why. 😕

Here's the full error information:
OpenSSL::SSL::SSLError: SSL SSLContext not available
  uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/net/http.rb:1018:in `connect'
  uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/net/http.rb:934:in `do_start'
  uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/net/http.rb:929:in `start'
  -:108:in `<main>'

You might have more luck using Mislav's SSL doctor.rb script. You can get it here:
https://github.com/mislav/ssl-tools/blob/8b3dec4/doctor.rb
Read more about the script and how to use it in this blog post:
https://mislav.net/2013/07/ruby-openssl/
[root@puppet ~]# rpm -qa |grep puppetserver
puppetserver-7.9.5-1.el8.noarch

```

**Comment by smortex** [comment]
Hum, the EL8 package seems broken in respect to SSL :unamused:…

```sh-session
# rpm -qa |grep puppetserver
puppetserver-7.9.5-1.el7.noarch
# curl -Lks 'https://git.io/rg-ssl'|/opt/puppetlabs/bin/puppetserver ruby
!!! WARNING !!!
Ruby 2.6.8 has reached end-of-life, and is unsupported.
This script may not work.

Here's your Ruby and OpenSSL environment:

Ruby:           2.6.8p0 (2022-10-24 revision 67951) [java]
RubyGems:       3.2.33
Bundler:        2.2.33
Compiled with:  JRuby-OpenSSL 0.14.0
Loaded version: JRuby-OpenSSL 0.14.0
SSL_CERT_FILE:  /etc/pki/tls/certs/ca-bundle.crt
SSL_CERT_DIR:   /etc/ssl/certs

With that out of the way, let's see if you can connect to rubygems.org...

Bundler connection to rubygems.org:       success ✅
RubyGems connection to rubygems.org:      success ✅
Ruby net/http connection to rubygems.org: success ✅

Hooray! This Ruby can connect to rubygems.org. You are all set to use Bundler and RubyGems. 👌
```

**Comment by huckabeec** [comment]
I'm going to guess it is FIPS - this server is in FIPS mode, which has a direct impact on anything ssl related.

**Comment by huckabeec** [comment]
`Net::LDAP::Error: FIPS mode: only SunJSSE KeyManagers may be used`

After writing a small test program for net-ldap in Ruby.  I've seen that error before with Java and it is fixed there by disabling FIPS mode for that one process but I don't know (yet) how to do that in this case.

**Comment by ekohl** [comment]
From https://github.com/theforeman/puppet-candlepin/commit/2fc73f033d2bb505c9e3502b8992e11323ec5287 I know we use `-Dcom.redhat.fips=false` for Candlepin, but no idea how that would work here.

**Comment by huckabeec** [comment]
Changing `security.useSystemPropertiesFile=true` to `security.useSystemPropertiesFile=false` in the java.security file for the local JDK fixes it.

There is probably some set of edits that would be made to the `/etc/crypto-policies/back-ends/java.config` file that re-enables a missing algorithm but I'm not wading into that mess.

This also fixes the gem install mess I was getting.  Now back to actually testing this module again.  Hopefully this is useful info for the next poor soul.

**Comment by huckabeec** [comment]
```
# /opt/puppetlabs/bin/puppetserver gem list |grep ldap
net-ldap (0.17.1)
```
So gem is installed for the puppetserver and not the agent, so what is the best way to test this as what I was doing before was obviously wrong ?

**Comment by smortex** [comment]
I only lurked the module and got my mind blown up, but never used it. I guess that would be something like that:

```puppet
ldapquery('(&(objectClass=inetOrgPerson)(...))', ['uid', 'uidNumber', ...]).each |$ldapuser| {
  user { $ldapuser['uid']:
    ensure => present,
    uid    => $ldapuser['uidNumber'],
    # ...
  }
}
```

The README documents each part, but maybe adding a "full real-world example" at the end can be a good addition :wink: ?

**Comment by huckabeec** [comment]
> The README documents each part, but maybe adding a "full real-world example" at the end can be a good addition 😉 ?

That's a reoccurring issue with most of the modules listed on the forge and elsewhere.  They have 'example' snippets of code, but each really needs a tutorial type example that shows how the module should be used, where the code goes, best practices, etc.  Too many give a one liner example and little if anything else.

**Comment by huckabeec** [comment]
`Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Function Call, uninitialized constant Net::LDAP::LdapError`

So something is wrong but it can't tell me what's wrong because of a missing constant.

**Comment by smortex** [comment]
Hum… https://github.com/ruby-ldap/ruby-net-ldap/commit/94b9d4d0443b8be37b91fcb5d6a31a9589cca41a

Maybe as a workaround remove the version you installed and install a version before 0.17.0:

```sh-session
# puppetserver gem uninstall net-ldap
# puppetserver gem search --all net-ldap --all
[...] --all
net-ldap (0.17.1, 0.17.0, 0.16.3, 0.16.2, 0.16.1, 0.16.0, 0.15.0, 0.14.0, 0.13.0, 0.12.1, 0.12.0, 0.11, 0.10.1, 0.10.0, 0.9.0, 0.8.0, 0.7.0, 0.6.1, 0.6.0, 0.5.1, 0.3.1, 0.3.0, 0.2.2, 0.2.1, 0.2, 0.1.1, 0.1.0, 0.0.5)
[...]
# puppetserver gem install -v 0.16.3 net-ldap
```

Restart the PuppetServer and retry.  I think that we can update this module to use `Net::LDAP::Error` instead to fix that, I may give it a look tomorrow.

**Comment by huckabeec** [comment]
After applying the fix from #38 and restarting everything, I've got a working config with FIPS and net-ldap 0.17.1.

Thanks to everyone for the help.
